lib/sgtn-client/api/translation.rb

```diff
@@ -181,27 +181,45 @@ def self.single_load(component, locale)
       end
     end
 
-    @@refresh_threads_lock = Mutex.new
-    @@refresh_threads = {}
+    @@refresh_semaphores = {}
+    @@refresh_semaphores_lock = Mutex.new
     def self.single_refresh(component, locale)
       cache_key = SgtnClient::CacheUtil.get_cachekey(component, locale)
-      @@refresh_threads_lock.synchronize do
```

Comment:
A thread locks @@refresh_threads_lock for entire application. It's better to lock only on the bundle being modified so that other thread working on other bundles are not blocked.

Reply:
There are several ways to solve this. one is in this PR. You may use this, or modify your code differently.

Reply:
You can also try locking on the individual thread. Let me make another PR for this option.

Reply:
```
            thread = Thread.new do
              items = load_and_compare_source(component, locale)
              SgtnClient::CacheUtil.write_cache(cache_key, items) if items&.empty? == false
              items
            end
```

Reply:
@@refresh_threads_lock is only locked to start a new thread.

Reply:
> Other threads are not blocked because a separate thread is started to update a bunde in background.
Yes, that's another thing. We don't want to start a new thread all the time. For a page with 10k strings (exaggerating here to make a point), it will spawn 10k new threads. We can avoid this.

PR 51 avoids creating a new thread when items is nil.
PR 50 avoids creating a new thread when items is nil AND creates only ~1 thread when expired.

Reply:
> That's another thing we can improve--we don't want to start a new thread all the time. For a page with 10k strings (exaggerating here to make a point), it will spawn 10k new threads. We can avoid this.
For a bundle(component/locale), there will be only one thread at the same time. Check below code, when a thread to refresh is running, then won't start other threads.
```
        thread = @@refresh_threads[cache_key]
        if thread.nil? || thread.alive? == false 
```